scripts/build_test_history.py

```diff
@@ -0,0 +1,447 @@
+    bucket = storage_client.bucket(BUCKET_NAME)
+
+    # Replace old revision number with new number.
+    blob = bucket.blob(f'{browser_name}_recent_statuses.json')
+    blob.upload_from_string(json.dumps(new_statuses))
+
+
+def _populate_previous_statuses(browser_name: str) -> dict:
+    """Create a dict with the most recent test statuses seen for browser."""
+    if SHOULD_GENERATE_NEW_STATUSES_JSON:
+        # Returning an empty dictionary of recent statuses will generate the
+        # initial recent statuses file and all of the first history entries.
+        return {}
+    # If the JSON file is not found, then an exception should be raised
+    # or the file should be generated, depending on the constant's value.
+    statuses_json_str = get_previous_statuses(browser_name)
+    if statuses_json_str is None:
+        # If this is not the first ever run for test statuses, then raise an
+        # exception if the JSON file was not found.
+        raise Exception(
+            f'Error obtaining recent statuses file for {browser_name}')
+
+    test_statuses = json.loads(statuses_json_str)
+    # Turn the list of recent statuses into a dictionary for quick referencing.
+    prev_test_statuses = {(t['test_name'], t['subtest_name']): t['status']
+                          for t in test_statuses}
+    return prev_test_statuses
+
+
+def should_process_run(run_metadata: MetadataDict) -> bool:
+    """Check if a run should be processed."""
+    # A run should be processed if no entities have been written for it.
+    test_entry = TestHistoryEntry.query(
+        TestHistoryEntry.RunID == run_metadata['id']).get()
+    return test_entry is None
+
+
+def process_runs(
+        runs_list: list[MetadataDict],
+        process_start_entity: MostRecentHistoryProcessed) -> None:
+    """Process each aligned run and update the
+    most recent processed date afterward."""
+    revisions_processed = {}
+    # Go through each aligned run.
+
+    start = time.time()
+    for run_metadata in runs_list:
+        browser_name = run_metadata['browser_name']
+        revision = run_metadata['full_revision_hash']
+
+        # Keep track of the runs that have been processed.
+        # The process start date entity is only updated once all aligned runs
+        # for a given revision are processed.
+        if revision not in revisions_processed:
+            revisions_processed[revision] = {
+                'chrome': False,
+                'edge': False,
+                'firefox': False,
+                'safari': False,
+            }
+
+        if should_process_run(run_metadata):
+            process_single_run(run_metadata)
+        else:
+            print('Run has already been processed')
+
+        revisions_processed[revision][browser_name] = True
+        # If all runs for this revision have been processed, we can update
+        # the most recently processed date to the run's start time.
+        if (revisions_processed[revision]['chrome'] and
```

Comment:
```suggestion
      if (revisions_processed[revision]['chrome'] and
```
This part should be outside of for loop

Reply:
There needs a else statement if a browser is missing, and we should call `update_recent_processed_date` regardless because it will always be missing

Reply:
Not that we have the check mentioned above (4 runs required), it shouldn't be possible for us to come across a condition when one of these runs is missing